AXL: protect AXL's transfer file lists in multi-threading transfers

rhaas80 opened the pull request:
Branch https://github.com/rhaas80/AXL/tree/rhaas/locks contains an implementation protecting against the race condition. Changes are: https://github.com/rhaas80/AXL/compare/rhaas/keyvalue...rhaas80:rhaas/locks

Implementation notes:

* this protects only `axl_kvtrees` but not the other global variables (that hold configuration options), since the latter are only accessed in the main thread so those do not share a race condition
* calling `AXL_Config` on a transfer that has already been dispatch is a race condition (since there is no lock on the individual `file_list` kvtrees) and not supported
* calling `AXL_Config` for the DEBUG settings is race condition (since `axl_debug` is not protected but is accessed from within the worker threads) while there are already dispatched transfers and will not be fixed

rhaas80 committed:
store pointer to file_list rather than getting it from global array

This removes the need to protect the global array. It is only accessed
by the main thread.

tonyhutter commented on src/axl.c:
```diff
@@ -1376,12 +1410,13 @@ int AXL_Free (int id)
 
 int AXL_Stop ()
 {
+    const unsigned int kvtrees_count = axl_kvtrees_count();
```
I believe you'll need to hold the lock for the the entirety of this function to prevent `axl_kvtree_count()` from changing

rhaas80 replied:
Good point. For `AXL_Stop` the issue may be bit more complex.  I think if `count` changes (keeping in mind that it can only increase) during this function I am in deep trouble to begin with since there would be active transfers left after `AXL_Stop`. On the plus side tight now `count` is only modified by the main thread, so there is not chance of it changing. If I wanted to make `AXL_Stop` thread safe, I would need to record a "stopped" flag to prevent new transfers from being created once `AXL_Stop` starts executing.

rhaas80 replied:
Looking at this I may be able to avoid the lock `axl_kvtrees` completely. The only reason it is there is that he pthreads only record the transfer ID integer and thus have to access the array to get the `file_list` pointer. However if one instead stored the `file_list` pointer (which never changes) in the `pdata` structure (in addition to the transfer ID) then one does not have multiple threads accessing `axl_kvtrees` and no lock is needed.

rhaas80 replied:
I made the change to copy `file_list` in ad3e3c4c59fcbc09f3cc4061e741b9f77b877f46 which avoids this issue altogether.

tonyhutter commented on src/axl.c:
```diff
@@ -490,9 +523,10 @@ static kvtree* AXL_Config_Get()
     /* this should be protected by a mutex_lock to prevent issues with
      * realloc() moving memory when growing axl_kvtrees, but no one else
      * does ... */
```
You many need to hold the lock here before iterating over the all the kvtrees, as the count could change.

tonyhutter commented on src/axl.c:
```diff
@@ -490,9 +523,10 @@ static kvtree* AXL_Config_Get()
     /* this should be protected by a mutex_lock to prevent issues with
      * realloc() moving memory when growing axl_kvtrees, but no one else
      * does ... */
+    const int kvtrees_count = axl_kvtrees_count();
```
You many need to hold the lock here before iterating over the all the kvtrees, as the count could change.

rhaas80 replied:
Not sure if `count` matters. Since it only ever increases, I might be missing some transfers in the report (and it would only be changed from within the main thread so there is no race condition as long as only the main thread calls `AXL_Config`). The transfers themselves (`file_list`) are currently only changing from within the main thread. 

I have to admit I am weary of keeping the lock any longer than I need to and in particular calling any function while holding the lock (just in case those functions ever end up trying to get the lock as well and the mutex being of the lock-only-once type).

rhaas80 replied:
uh, actually. I *do* need to hold the lock since the array pointer becomes invalid,l which i guess is what you had in mind.

rhaas80 replied:
> uh, actually. I _do_ need to hold the lock since the array pointer becomes invalid,l which i guess is what you had in mind.

second uh. Maybe not in the end. As long as I am using the `avl_kvtree_get` function the change of pointer of the array does not affect me (and once more it should not change as long as I am on the main thread).

rhaas80 replied:
The latest version no longer protects `axl_kvtrees` and instead copies the required entry (the `file_list` pointer) when a new transfer is created. This means this issue no longer applies.

tonyhutter commented on src/axl.c:
```diff
@@ -69,11 +69,43 @@ int axl_copy_metadata;
 
-static pthread_mutex_t id_lock = PTHREAD_MUTEX_INITIALIZER;
+/* lock and return all the AXL_Create'd kvtree pointers */
+static axl_kvtrees_t* axl_kvtrees_obtain(void)
```
I would rename these functions to `axl_kvtrees_lock()`/`axl_kvtrees_unlock()` to match the pthreads conventions.

rhaas80 replied:
That is actually the reason why I do not name them lock/unlock. They do more: they return a pointer to the object (obtain it).

rhaas80 replied:
The latest version no longer protects `axl_kvtrees` and instead copies the required entry (the `file_list` pointer) when a new transfer is created. This removes the functions and any issue about their name.

rhaas80 commented:
I changed the code so that no locking of `axl_kvtrees` is required anymore. This assumes, like most other code in AXL, that AXL's user facing functions are only called by a single thread.

tonyhutter commented:
Looks like your AXL_Config code was merged. Can you make this PR to be against the master branch?

rhaas80 commented:
> Looks like your AXL_Config code was merged. Can you make this PR to be against the master branch?

I will try. I have to find out how to change the base branch in GitHub to a branch in a different repository. Since pull requests seem to "belong" to the repository that "owns" the base branch this may require me creating a new pull request.

rhaas80 commented:
I created a new pull request in the the VeloC components repo in https://github.com/ECP-VeloC/AXL/pull/87

gonsie committed:
coc and contrib

adammoody committed:
drop -Werror from default build flags

adammoody committed:
drop macro mangling for axl interface

tonyhutter committed:
Record metadata to state_file by default

No matter what, we need to do a stat call on each file to get the file's
size.  This stat call also gives us all the file's metadata (permissions,
UID, GID, creation times) "for free".  This patch encodes the size and all
metadata by default into the state_file.  This is needed for finalizing
transfers after a burst buffer transfer has finished, and the source files
are no longer available.

This change does not affect AXL_COPY_METADATA.  It must still be set in
order to apply the metadata to the file after it has been transferred.

tonyhutter committed:
Don't load the BB API when doing a non-BBAPI transfer

Previously, we would blindly call axl_async_init_bbapi() and
axl_async_finalize_bbapi() in AXL_Init/AXL_Finalize if we detected the
BBAPI headers.  This became problematic on the IBM login nodes since
they had the BBAPI headers, but weren't running the client.

This patch changes the behavior to only load the BBAPI library if
you're doing a BBAPI transfer.

Fixes: #70
